### ISSUE TITLE
Allow superusers to edit "publisher id" field in Settings

### ISSUE THREAD

**Issue by emmajclegg** [younginnovations/iatipublisher#1298]
Can IATI Publisher superusers (i.e. those of us that work on the IATI Support helpdesk) be allowed to edit an organisation's "publisher ID" in the Settings menu?

We agree normal users should not be able to do this, but there will be times where it makes sense for IATI Support to be able to (e.g. if the publisher ID is not intuitive and the organisation wants it updated). We understand the change will need to happen first on the IATI Registry, and the API token updated after.

![image](https://github.com/younginnovations/iatipublisher/assets/137316374/c67f50db-e460-43b7-b523-89e700dd4871)

**Comment by emmajclegg** [comment]
@praweshsth - I found this old issue as the same question has come up again (I need to change an organisation's publisher ID in IATI Publisher because they want to change it on the Registry). 

Was this fix ever implemented? I still can't edit an organisation's Publisher ID field while logged in as an IATI Admin.

**Comment by praweshsth** [comment]
@emmajclegg we haven't worked on it.

**Comment by emmajclegg** [comment]
No worries - I'll reopen this and add it to our contract monitoring spreadsheet. It'll be lower priority than the existing issues on there.

**Comment by emmajclegg** [comment]
Hi @praweshsth - I see that we've estimated approx. 2.5 days work for this on the Contract Monitoring spreadsheet. Can we confirm the workflow steps between us as I want to check my understanding of how it should work too.

My understanding of workflow:
1. An IATI admin (i.e. a member of the IATI Support team) logs onto the IATI Registry, changes the "publisher ID" for the required IATI publisher account and saves changes.
2. The IATI admin logs into IATI Publisher, proxies the organisation's account and updates their publisher ID to match what's now on the Registry.
3. IATI Publisher unpublishes any data files that have already been published by the organisation
4. The API key connection between Publisher and the IATI Registry for this organisation's account is updated 
5. IATI Publisher republishes the organisation's data files, from the updated publisher account

Do we agree that steps 3-5 are necessary? (I'm taking information from this [IATI website guidance](https://iatistandard.org/en/guidance/publishing-data/registering-and-managing-your-organisation-account/updating-your-organisation-details/) but have not gone through the process before). How many of these steps can we automate would be the second question?

Happy to discuss. cc'ing @IsabelBirds for info, in case you've anything to add on the workflow from experience

**Comment by A4family** [comment]
Hi @emmajclegg ,

# Proposed Workflow:

1. An IATI admin, a member of the IATI Support team, logs into the IATI Registry and updates the "publisher ID" for the required IATI publisher account, then saves the changes. ***( SAME )***
2. The IATI admin accesses the IATI Publisher directly (without proxying the organization's account) and uses the edit option in the organization's list to modify the publisher's ID.
3. We avoid letting the IATI admin update the publisher's Id by proxying the organization because doing so would change the admin's role to another user of the organization, who lacks permission to modify the publisher's Id.

# Affected Areas:
>[!note]
>1. Organisation's XML file
>2. Activity individual XML file
>3. Merged activity XML file
>4. Update the link in the IATI registry of merged activity XML file

The publisher ID primarily serves as the basis for naming individual XML files for activities, merged XML files for activities, and the organization's XML file stored in the S3 bucket. Changing the publisher's ID also requires updating the names of the organization's related XML files in S3 and updating the link to the merged published XML file in the IATI Registry. 

In response to your concern, steps 3 and 5 are not necessary because manually handling changes for hundreds of already published activities within an organization would not be possible. So, writing a script is necessary to automate the process of renaming related XML files and updating the links in the IATI registry.

cc @IsabelBirds @praweshsth

**Comment by emmajclegg** [comment]
Thanks for the extra detail @A4family - I'm happy to start this work based on what you've described. I wasn't sure what the impact of changing a publisher ID has on the Registry API token associated with the organisation's account.

@praweshsth - I mentioned today that there's one publisher who's been waiting to change their publisher ID for a while. I would prefer to concentrate on the long term fix we're discussing in this issue but, if it's looking likely to take more than 2 weeks, we might need a workaround. I'm happy to see where we are next week (week beginning 25th March) and review then.

**Comment by sarinasindurakar** [comment]
@PG-Momik 

- [ ]  The settings page fails to load, displaying only a blank white screen
![image](https://github.com/younginnovations/iatipublisher/assets/65848732/5d13b3ab-b9f6-492c-aedf-1a54d657b3a6)

**Comment by PG-Momik** [comment]
@emmajclegg . This has been deployed to staging.

> [!NOTE]
Note that changing `Publisher ID` on **Publisher** only works well, if the **IATI Archiver** has completed its work on the IATI Registry (takes 30-40 seconds) after updating publisher info on **IATI registry**.
![Image](https://github.com/younginnovations/iatipublisher/assets/57512059/e8794ef8-8746-4f91-af50-4fe262b25b8e)

**Comment by emmajclegg** [comment]
Thanks @PG-Momik. Can I check the expected workflow here to make sure I'm following it correctly:

**1) An IATI admin logs into the IATI Registry and updates the "publisher ID" for the required IATI publisher account, then saves the changes.** 

Done - I changed my publisher ID from _emma-test2-publisher_ to _emma-test2-publisher-changed_ and saved.
![image](https://github.com/younginnovations/iatipublisher/assets/137316374/a50d5697-4015-44d0-a17c-ccc09360ce76)


**2. An IATI admin runs the IATI Archiver to update the datasets currently published**

I've manually triggered the IATI Archiver ([see activity stream](https://staging.iatiregistry.org/publisher/activity/emma-test2-publisher-changed/0)) but it causes a metadata error on the activity file:

![image](https://github.com/younginnovations/iatipublisher/assets/137316374/e008e11c-f772-48fa-90d1-e34f8f3db9aa)


**3. The IATI admin accesses IATI Publisher directly (without proxying the organization's account) and uses the edit option in the organization's list to modify the publisher's ID.**
I don't see a way to edit an organisation from the Organisation List page without proxying their account? 

![image](https://github.com/younginnovations/iatipublisher/assets/137316374/97208475-a348-4305-856d-98f063e71aa2)

When proxied into the account, I see the "publisher is inactive" alert on the Activity Data page, and publishing settings are now blank:

![image](https://github.com/younginnovations/iatipublisher/assets/137316374/7c8a2746-d316-4c50-bf1a-034fd9b18c09)


**4. The IATI Admin populates the new publisher ID, before asking the user to generate and update their API token?**
(I thought the user would be able to avoid needing to update their API token, but I'm not sure how this is intended to work)

When I updated the publisher ID and created a new API token for the [associated user,](https://staging.iatiregistry.org/user/emma-test2-user/api-tokens) I got the error message "Publisher ID or API token incorrect":
![image](https://github.com/younginnovations/iatipublisher/assets/137316374/9e9ab99c-4abf-4e92-ab28-11335ca21db1)



Any clarity on the workflow welcome, then happy to retry.

**Comment by PG-Momik** [comment]
@emmajclegg 
> I don't see a way to edit an organisation from the Organisation List page without proxying their account?

After an internal discussion, we decided on not letting IATI Admins edit publisher id via an edit button on the organization listing page, and only needing superadmins/iati-admins to proxy to edit the publisher id. The edit functionality via  popup model on clicking the edit button would indicate that the  organisation data is being edited, but in actuality only the publisher id would be indicated. On a db level publisher id is stored on `settings` rather than `organisation`. Implementing the popup modal to edit the publisher id , resulted in a popup modal with only 1 field, which seemed completely unnecessary. So we decided to remove it. 

Apologies for not mentioning this decision.

**Comment by PG-Momik** [comment]
@emmajclegg 

## Workflow Summary

The ideal workflow for this issue:
1. User changes publisher in IATI Registry.
2. User contacts superadmin/iati admin.
3. Iati-admin changes publisher ID. 
    - Iati-admin logs into IATI Publisher
    - Iati-admin proxies into organisation .
    - Iati-admin navigates to settings.
    - *(On page load, Iati-admin sees validation error on `Publisher ID` field. No correct/incorrect/pending tag.)*
    - Iati-admin edits token.  (`Publisher ID` field is only editable for Superadmin and Iati-admin)
    - Iati-admin saves changes.
    - *(Validation error on `Publisher ID` field disappears, correct/incorrect/pending tag shows up. Dataset url in IATI registry changes)*
   
---

> ...., and publishing settings are now blank:

There was a slight implementation issue on the backend ,that didnt handle initial page load with incorrect `Publisher ID` . This has been resolved. Other changes added are:
- Show validation error on `Publisher ID` field incase publisher id does not exist on registry. 
- Show API token, but not show correct/incorrect/pending tag if `Publisher ID` does not exist on registry.
- Prevent form save if `Publisher ID` incorrect.

> 4. The IATI Admin populates the new publisher ID, before asking the user to generate and update their API token? 

We were not able to recreate this issue.  This could have been resolved because of other changes along the way. Please attach a screen recording if this issue persists.

**Comment by emmajclegg** [comment]
Thanks @PG-Momik - I've got this working as expected now in IATI Publisher.

On the Registry staging site, it looks like the activity file url gets updated with the new publisher ID, but the organisation file doesn't?

_Example_: I've changed the publisher ID for my test publisher account (https://staging.iatiregistry.org/publisher/emma-test2-publisher-reverted) from "emma-test2-publisher" to "emma-test2-publisher-reverted". I updated the Settings in IATI Publisher.

On the Registry staging site, the IATI Archiver has run on both the published activity and organisation files, however metadata appears missing for the organisation file:


![Image](https://github.com/younginnovations/iatipublisher/assets/137316374/51721a05-da1f-4dd5-a992-e1e52430cf2e)

The organisation file url still contains the old publisher ID: https://iatipublisher-staging.s3.amazonaws.com/organizationXmlFiles/emma-test2-publisher-organisation.xml

**Comment by sarinasindurakar** [comment]
@PG-Momik 
If the organization files are in a republished state and either the superadmin or IATI Admin changes the publisher ID, the organization dataset does not get updated with the new publisher ID in the registry. The user are also unable to publish the organisation data. Please look attached video to replicate the steps :

[Untitled_ Apr 03,2024 1_07 PM.webm](https://github.com/younginnovations/iatipublisher/assets/65848732/da5b2842-917e-46f0-a613-64335fb3e971)

**Comment by emmajclegg** [comment]
@PG-Momik - I've just tested changing _publisher ID_ again on two test organisation accounts.

Everything worked as expected the first time I changed the publisher ID for organisation [Emma's test publisher v.2](https://staging.iatiregistry.org/publisher/emma-test2-publisher-new3) (I changed this in the Registry first, then in IATI Publisher settings. The Registry activity and organisation files urls were both updated as expected to include the new publisher ID string.)

However, changing the publisher ID a second and third time for the same publisher account, and following the same steps, has left me with duplicate activity and organisation files on the Registry:



![Image](https://github.com/younginnovations/iatipublisher/assets/137316374/b598de3f-9e9f-4518-9d65-765c3da5a37b)







It's unlikely we'll ever be changing publisher IDs multiple times like this, but I wanted to check what would happen. Thinking it might be related to changing the publisher ID multiple times in a row, I then tried on a different publisher account ("[My example organisation](https://staging.iatiregistry.org/publisher/emma-test5-publisher-changed)") and the same duplication happened:



![Image](https://github.com/younginnovations/iatipublisher/assets/137316374/e57b310e-36ad-4692-b03a-533e56d44562)


This was caused by changing the publisher ID from _emma-test5-publisher_ to _emma-test5-publisher-changed_.

Any idea what's causing this file duplication?

**Comment by PG-Momik** [comment]
@emmajclegg 

[As previously mentioned here.](https://github.com/younginnovations/iatipublisher/issues/1298#issuecomment-2022548517) Changing Publisher Id on IATI Publisher will only work well if the IATI Archiver completes updating the datasets in IATI Registry. Upon changing publisher Id on IATI Registry please wait a few moments for IATI Archiver to finish its work. The activity stream should look like this after 35 seconds of updating publisher id in registry:




![Image](https://github.com/younginnovations/iatipublisher/assets/57512059/d42ffb63-de42-4f00-b465-4710b96be791)



Then you can proceed to change publisher id in IATI Publisher.

---


Dataset with errors can be removed from registry [from manage button >> dataset tab](https://staging.iatiregistry.org/publisher/bulk_process/emma-test2-publisher-new5)

![Image](https://github.com/younginnovations/iatipublisher/assets/57512059/e5327e7b-7b95-466c-859c-afd412c1c479)


---

cc : @sarinasindurakar

**Comment by emmajclegg** [comment]
Ok @PG-Momik, happy to move on to deploying this. Thanks
